org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ReadManager.java

```diff
@@ -32,19 +35,20 @@ public class ReadManager {
 	 * (first in first out).
 	 */
 	private static int THREAD_COUNT = Math.max(0, Math.min(CACHE_SIZE, Runtime.getRuntime().availableProcessors() - 2));
-	private static final ExecutorService executor = THREAD_COUNT <= 0 ? null
-			: Executors.newFixedThreadPool(THREAD_COUNT, r -> {
-				Thread t = new Thread(r, "Compiler Source File Reader"); //$NON-NLS-1$
-				t.setDaemon(true);
-				return t;
-			});
+	private static final ExecutorService READER_SERVICE = THREAD_COUNT <= 0 ? null
+			: new ThreadPoolExecutor(THREAD_COUNT, THREAD_COUNT, /* keepAliveTime */ 5, TimeUnit.MINUTES,
```

Comment:
I believe in order to stop idling threads, it should be something like `new ThreadPoolExecutor(1, THREAD_COUNT)`. At least I don't see threads disappearing with the current patch state.

Same for BatchImageBuilder

Reply:
Unfortunately that API seems to be not intuitive - corePoolSize=1 would never create more then 1 thread with an unbounded queue.
I'll take this workaround:
https://stackoverflow.com/a/19528622